Add database

Anarmard opened the pull request:
Не могу понять/увидеть, почему не создается фильм. Create Film. 

По User все тесты проходят.

Anarmard committed:
Добавили базу данных, но программа не запускается

Anarmard committed:
Все тесты по User успешно проходят
Проблема с созданием FILM

alexache commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,14 @@
+@ToString
+//@EqualsAndHashCode(of = "ratingMpaId")
+public class Mpa {
+    private Long ratingMpaId;
```
Поля должны называться id и name иначе тысты постман не пройдут

Anarmard replied:
Поправил. Спасибо! А как ты узнаешь, как должны называться поля для успешного прохождения тестов в постмане?Это надо смотреть в постмане или в идеи? Если в постмане, то это на вкладке Tests?(pm.expect(jsonData.**id**).to.eql(1))

alexache replied:
Я смотрел в постмане, но можно и в самом файле, это не принципиально где смотреть, там видно, что конкретно ожидается, и вот это вот "jsonData.id" и говорит об ожидаемом имени

alexache commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,13 @@
+@ToString
+@EqualsAndHashCode(of = "genreId")
+public class Genre {
+    private Long genreId;
```
Ёти поля тоже должны называться id и name дляп прохождения тестов

Anarmard replied:
Сделал

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,221 @@
+
+    // создание фильма
+    @Override
+    public Optional<Film> createFilm(Film film) {
```
Не ояень логияно возвращать опшенл на создание сущности. Если не полцчится создать должна быть ошибка, а не просто пустоц опшенл

Anarmard replied:
Убрал Optional в методе Create

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,159 @@
+                .usingGeneratedKeyColumns("USER_ID");
+
+        log.debug("Добавлен пользователь {}", user);
+        return getUserByID(simpleJdbcInsert.executeAndReturnKey(toMapUser(user)).longValue());
```
Тоже довольно странная конструкция с добавление и сразу поиском по БД

Anarmard replied:
Поправил

alexache commented on src/main/resources/application.properties:
```diff
@@ -1 +1,7 @@
+spring.sql.init.mode=always
+# в jdbc-url укажите, что данные нужно сохранять в файл
+# spring.datasource.url=jdbc:h2:file:./db/filmorate
+spring.datasource.url=jdbc:h2:mem:filmorate
```
почему только в памяти а не в файле?

Anarmard replied:
Если поставить сохранение в файл и запустить приложение "FilmorateApplication", то первый раз тесты постмана проходят как ожидаешь - что-то работает, что-то нет. Но вот если остановить приложение и снова его запустить и снова запустить тесты, то даже успешные тесты в 1-й попытке становятся неуспешными. 
Как я понял из slack, это из-за того, что в файле filmorate.mv.db при первом прохождении тестов сохраняются данные, которые и сбивают последующие запуски тестов. Там же в слаке прочитал, что можно поменять на память, чтобы каждый раз после прогона тестов не удалять вручную файл filmorate.mv.db

alexache replied:
Для локальной проверки можно не использовать файловое хранение, а вот для пуша в репозиторий лучше возвращать файловое хранение, в идеале это должны быть разные профили

Anarmard replied:
Понял, когда буду делать пул реквест верну настройку обратно на сохранение в файле.

alexache commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -1,13 +1,32 @@
-	}
+	public void testFindUserById() {
 
+		Optional<User> userOptional = userStorage.getUserByID(1L);
```
а откуда ожидается пользователь с таким айди на старте пустоц системы? Из-за этого падающего теста ломается сборка и соотвественно не видно остальных постман тестов

Anarmard replied:
Александр, согласен с твоим комментарием!! Поправил - добавил создание фильма с id 1. Спасибо тебе!

Но вот наверное не до конца понимаю, что такое "ломается сборка". Даже с не работающим тестом FilmorateApplicationTest , приложение FilmorateApplication у меня почему-то запускался и запускается... И даже команда "mvn spring-boot:run" вроде срабатывает. "Неработающая сборка" - где можно увидеть, что сборка не работает?

alexache replied:
Возможно проблема в моей вольной терминалогии, Я имею в виду сборку в гитхабе при пуше коммитов, сейчас она красная и написано в ПР, что есть "All checks have failed" с указанием на то, где что-то пошло не так

Anarmard replied:
Ок, ясно!

alexache reviewed (changes requested):
Не ручаюсь, что нашел все проблемы, но комментарии должны помочь исправить прохождение билда и создание фильмов, пришлось локально подебажить

Anarmard commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -77,4 +92,19 @@ public void validateFilm (Film film) {
+
+    public void checkFilmId(Long filmId) {
+        Optional<Film> filmResponse = filmService.getFilmByID(filmId);
+        if (filmResponse.isEmpty()) {
```
Спасибо, поправил!

Anarmard commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -27,32 +33,41 @@ public Collection<Film> findAllFilms() {
 
     @GetMapping("/{id}") // получение фильма по ID
-    public Film getFilmByID(@PathVariable Long id) {
+    public Optional<Film> getFilmByID(@PathVariable Long id) {
```
Поправил, спасибо.

Anarmard commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        log.debug("Добавлен фильм {}", film);
+        return getFilmByID(filmId);
```
Да, Александр, это решило проблему!!! Спасибо огромное!

Anarmard committed:
Внес правки по возврату film, а не Optional.
Методом проб и ошибок, есть гипотеза, что тесты сыпятся из-за "getFilmByID", но ошибку найти не получается.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -96,6 +98,7 @@ public Optional<Film> getFilmByID(Long filmId) {
                     filmRow.getDate("RELEASE_DATE").toLocalDate(),
                     filmRow.getInt("DURATION"),
+                    filmRow.getInt("RATE"),
                     new Mpa(filmRow.getLong("RATING_MPA"), filmRow.getString("DESCRIPTION")),
```
Из-за этой строчки не проходит один из тестов постмана. Подсказка - у тебя description берется из таблицы фильма, а не рейтинга

Anarmard replied:
Аааа..... Александр, спасибо огромное! Я всё не мог понять, почему в процессе дебага на этой строке у меня происходила ошибка ... Теперь намного лучше! осталось пару тестов в постмане

alexache reviewed (changes requested):
Еще не проходит много тестов постмана

Anarmard committed:
Поправил название столбцов в SQL запросах
Обращать внимание на используемые столбцы, если названия столбцов совпадают из разных таблиц

Anarmard commented:
Постарался поправить ошибки на которые ты указал. Теперь все тесты в постмане проходят. Александр, спасибо за терпение!
